scipy/sparse/_construct.py

```diff
@@ -465,8 +465,18 @@ def _stack_along_minor_axis(blocks, axis):
     # Do the stacking
     indptr_list = [b.indptr for b in blocks]
     data_cat = np.concatenate([b.data for b in blocks])
-    idx_dtype = get_index_dtype(arrays=indptr_list,
-                                maxval=max(data_cat.size, constant_dim))
+
+    # Need to check if any indices/indptr, would be too large post-
+    # concatenation for np.int32:
+    # - The max value of indices is the output array's stacking-axis length - 1
+    # - The max value in indptr is the number of non-zero entries. This is
+    #   exceedingly unlikely to require int64, but is checked out of an
+    #   abundance of caution.
+    max_int32 = np.iinfo(np.int32).max
+    sum_dim = sum(b.shape[axis] for b in blocks)
+    nnz = sum(len(b.indices) for b in blocks)
+    needs_64bit = max(sum_dim - 1, nnz) > max_int32
+    idx_dtype = np.int64 if needs_64bit else np.int32
```

Comment:
I think it would be nice to continue using get_index_dtype here for consistency. You could replace this and the previous line with:

```python
idx_type = get_index_dtype(maxval=max(sum_dim - 1, nnz))
```